### DISCUSSION TITLE
Crop sowing / harvest scheduling problem, solver 'stuck'

### DISCUSSION THREAD

**Discussion by jddingemanse** [google/or-tools#3604] [CP-SAT questions] [answered]
I started this week, in Python, on working with constraints for a crop scheduling problem: different crops, with different crop cycles (after sowing, some number of months until yield and therefore profit), with constraints such as available water per month, difference in water use per crop, available land, and some months that crops cannot grow. The **objective** is to **maximize yield/profit** (hectare per crop * profit per hectare). However, the solver gets stuck after adding a particular set of constraints, and I would like to know whether that can be solved. _Apologies on beforehand, being new to ortools and constraint modelling, my code is probably unnecessary long and I might be adding an unnecessary amount of constraints. Also, I do not how to reproduce my problem in a smaller setting, so I share the full code below._

### Explanation of the code
The current code creates for eight crops (x1-x8) and 12 months a variable (x1m1 up to x8m12), representing hectare per crop per month, whose hectare*profit combined is to be maximized. Constraints are created for water- and land availability, and besides that constraints representing sowing, cropcycle and harvesting _(assumptions: if sown, the following months within cropcycle there can be no sowing; if sown, that month and the following months there must be crop present; if not sown within the same month or preceding months for lenght of cropcycle, there can be no crop)_.
Next to the cropvariables, also for every crop/month combination a BoolVariable (x1m1sow up to x8m12sow) is created that represents sowing (1), or not (0). This variable is used to set the above mentioned constraints.

To control that certain seasons, certain crops cannot grow, the dict `cropOffSeason` can be set as input. With an empty dictionary (`cropOffSeason = {}`), every crop can grow every month. With `cropOffSeason = {'x4':[9]}`, the constraint `x4m9==0` is added --> no hectare for crop x4 in month 9.

### The problem / questions
Now the problem arises when I start to use the `cropOffSeason` dictionary more intensively. With empty dict it solves under a second (result will be only yield for x4, as that is the one with highest profit). Turning one x4 month off (`cropOffSeason = {'x4':[9]'}`) also works fine, and for those months crop x8 is grown (next best yield). However, as soon as I add an OffSeason for 'x8' (so, `cropOffSeason = {'x4':[9],'x8':[9]`), the solver gets stuck (it keeps on calculating for multiple hours, I've not yet reached a result, and after multiple hours I gave up).

--> Any suggestions on what I am doing wrong, or why the solver gets 'stuck' after adding just the single x8m9==0 on top of the already many constraints?
--> Any suggestions on adding the cropOffSeason functionality in a proper way - or any suggestions regarding how the other parts are implemented currently (due to my limited knowledge of ortools I only use the .Add() and .OnlyEnforceIf(); possibly there are more efficient ways of implementing the constraints?)
Preferably suggestions do not include 'buy a better laptop' :-)

Hereby the code; _edit: below the code I have added the first part of the log_

```py
# Import packages
from ortools.sat.python import cp_model
import pandas as pd
import numpy as np

############################## USER SETTINGS ########################################

available_water_per_month = [1705540]*12 #List with water availability per month, 12 months
available_land = 1176 #Fixed total number of hectare

crops = ['x1','x2','x3','x4','x5','x6','x7','x8'] # List of 2digit cropnames
cropcycle = dict(zip(crops,[3,4,4,4,5,12,4,4])) # For every crop, a number of months from sowing to harvest
waterconstraint = np.array([321,317,271,298,322,582,262,352]) # For every crop, the relative water usage
profit = [74750,38711,93600,192805,90060,135919,62483,123744] # For every crop, the profit per hectare of yield

cropOffSeason = {'x4':[9]} # Optionally, for crops, which months they cannot grow <<< adding this one extensively 
                                          # results for me in a 'stuck' solver
# cropOffSeason = {'x4':[9],'x8':[9]} # <-this one gets my solver stuck.

######################### PREPARING THE SOLVER ############################

# from ortools.sat.python import cp_model
model = cp_model.CpModel()

# Create the variables: for 12 months * number of crops.
no_months = 12

var_upper_bound = 100000
months = str(list(range(1,no_months+1)))[1:-1].split(', ')
cropvars = {} # In this dict, all the variables representing hectare per crop per month
sowvars = {}  # In this dict, all the Boolean variables representing sowing (1) or not (0)
for crop in crops:
    for month in months:
        cropvar = crop+'m'+month
        sowvar = crop+'m'+month+'sow'
        cropvars[cropvar] = model.NewIntVar(0, var_upper_bound ,cropvar)
        sowvars[cropvar] = model.NewBoolVar(sowvar)

cropsDf = pd.DataFrame({'names':cropvars.keys(),'variables':cropvars.values()}).set_index('names') # A DataFrame for some metadata about the crops, variables and later on results

cropsDf['crop'] = cropsDf.index.str.slice(stop=2)
cropsDf['month'] = cropsDf.index.str.slice(start=2)
cropsDf['waterconstraint'] = cropsDf.crop.replace(crops,waterconstraint)
cropsDf['profit'] = cropsDf.crop.replace(crops,profit)
cropsDf['cropcycle'] = cropsDf.crop.replace(crops,cropcycle.values())
cropsDf['monthInt'] = cropsDf.month.str.slice(1).astype(int)

# Create water constraints: per month the sum of crops hectare * their water usage should be <= the available water
for i, month in enumerate(months):
    monthname = 'm'+month
    multiply = cropsDf.variables*cropsDf.waterconstraint
    monthmultiply = multiply[cropsDf.index.str.slice(2)==monthname]
    model.Add(sum(monthmultiply)<= available_water_per_month[i])

# Create land constraints: Per month the sum of all crophectare should be <= the available land
for month in months:
    monthname = 'm'+month
    monthvars = cropsDf.variables[cropsDf.month==monthname]
    model.Add(monthvars.sum() <= available_land)    

# Certain crops cannot grow during certain months: simple constraints of hectare should be 0 based on the dictionary cropOffSeason
if len(cropOffSeason)>0:
    for crop in cropOffSeason.keys():
        yrmonths = np.array(cropOffSeason[crop])
        monthsOff = yrmonths.astype(int)
        for month in monthsOff:
            variable = cropsDf.variables[(cropsDf.monthInt==month)&(cropsDf.crop==crop)][0]
            model.Add(variable==0)

# Sowing constraints
#### Depending on cropcycle: if crop is sown (sowvars[crop]==True), the cropvars of the following months for duration of cropcycle should be equal
#### Also, then all next sowvars[crop] should be false (no sowing of a crop when it is already busy with a cropcycle)
for cropvar,sowvar in sowvars.items():          # Take all cropvars and sowvars one by one
    crop = cropvar[:2]                          # Take cropname
    cycle = cropcycle[crop]                     # Take cycle that matches cropname
    month = cropvar[2:]                         # Take month from cropvar
    monthno = int(month[1:])                    # Take month as integer
    for i in range(monthno+1,monthno+cycle):    # Go over all months succeeding the cropvar month 
        if i > no_months:                             # Break if outside of 12 month range
            i -= 12
        varname = crop+'m'+str(i)               # Create cropvar names succeeding the cropvar 
        model.Add(cropvars[cropvar]==cropvars[varname]).OnlyEnforceIf(sowvar)   # Succeeding months must be                                                                     
                                                                                # equal to cropvar month when
                                                                                # sowvar of original month 
                                                                                # is true.
        model.Add(sowvars[varname]==0).OnlyEnforceIf(sowvar)    # Succeeding sowvars must be False because 
                                                                # you cannot sow again on same ground until
                                                                # cropcycle is finished.

#### A cropvar must be zero, if within it's preceding cropcycle all sowvariables are zero or own sowvar is zero 
#### (there can be no hectare of crop if there has been no sowing that month or any of the previous months for duration of the cropcycle)
for cropvar, sowvar in sowvars.items():         # Go over all cropvars
    crop = cropvar[:2]                          # Take cropname
    cycle = cropcycle[crop]                     # Take cycle that matches cropname
    month = cropvar[2:]                         # Take month from cropvar
    monthno = int(month[1:])                    # Take month as integer
    
    constraint = f'model.Add(cropvars[\'{cropvar}\']==0)'       # Start basic constraint string

    for i in list(range(monthno-cycle+1, monthno+1))[::-1]:   # For the length of cropcycle, go over all preceding cropvars (limited by size of year) 
        if i <= 0: 
            i += 12
        varname = crop + 'm' + str(i)
        constraint = constraint + f'.OnlyEnforceIf(sowvars[\'{varname}\'].Not())'   
                # Add OnlyEnforceIf(sowvars[preceding_cropvar].Not()) to constraint string
    exec(constraint)    # Execute constraint string

#### A month cannot be zero, if the corresponding sowvariable is one (there must be some hectare of crop if that month there has been sowing)
for cropvar, sowvar in sowvars.items():
    model.Add(cropvars[cropvar]>0).OnlyEnforceIf(sowvar)

# solve it
# to take crop cycles into account (per cycle, only one yield), divide profit by cropcycle
profitweight = cropsDf.profit//cropsDf.cropcycle
model.Maximize(sum(cropsDf.variables*profitweight))
solver = cp_model.CpSolver()

solver.parameters.log_search_progress = True

status = solver.Solve(model)

for name,variable in zip(cropsDf.index,cropsDf.variables):
    cropsDf.loc[name,'hectare'] = solver.Value(variable)
    cropsDf.loc[name,'result'] = str((solver.Value(sowvars[name]),solver.Value(variable)))

cropsResult = cropsDf.pivot(index='monthInt',columns='crop',values='result')
print(cropsResult)
```
### Edit: added log output
With `solver.parameters.log_search_progress = True` added: I'm not sure what part of the log to add here, but hereby txt files, one for `cropOffSeason = {}` that gives a solution ([log_solution.txt](https://github.com/google/or-tools/files/10303438/log_solution.txt)), and one for `cropOffSeason = {'x4':[3,4,9]}` with the screen output for when my console gets 'stuck' ([log_stuck.txt](https://github.com/google/or-tools/files/10303453/log_stuck.txt)).

**Comment by lperron** [comment, marked as answer]
First, you should set 

    solver.parameters.log_search_progress = True 

and show the start of log

**Reply by jddingemanse** [reply]
I have updated my post with at the bottom text files of the log - I'm not sure which part of the log can be considered start. And already thanks, this one at least shows to me that my computer is busy while it is stuck, it seems (it regularly gives an update) - and that the presolver works fine, search starts within a second, only after that it keeps on searching. Should I set some limit and accept whatever is found by then?

**Reply by lperron** [reply]
yes.

With the extra constraints, your problem just became a lot harder.

**Reply by jddingemanse** [reply]
Ok, clear. I found the `solver.parameters.max_time_in_seconds`; at least I get a solution now. Thanks a lot for your directions! 

With the risk of asking stupid questions, two follow-up questions:
- Is there anything to say about the quality of the solution found versus max_time_in_seconds // any rationale for picking at least x seconds // big difference between 10 seconds or 2 hours?
- Is solving the harder problem related to CPU - if I put my university's HPC to work, is it more likely that the solver will find a best solution within a given time for the harder problems // does `solver.parameters.max_time_in_seconds = 10` on one laptop result in evaluating much more solutions than on another laptop?

**Comment by lperron** [comment]
the solver will continue searching, until it hits the time limit.
The biggest gain will be using more cores. We tune the solver for 16
workers.

Now all these problems are NP-Hard, thus exponential, and finding solutions
boils down to luck.

So just experiment to get a feel for the problem.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00



Le mar. 27 déc. 2022 à 15:20, jddingemanse ***@***.***> a
écrit :
